File: captainhcg/go/148-sort-list.go
func sortList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	dummy, middle := &ListNode{}, findMiddle(head)
	pre := dummy
	n1, n2 := sortList(head), sortList(middle)
	for n1 != nil && n2 != nil {
		if n1.Val <= n2.Val {
			pre.Next, pre, n1 = n1, n1, n1.Next
		} else {
			pre.Next, pre, n2 = n2, n2, n2.Next
		}
	}
	for n1 != nil {
		pre.Next, pre, n1 = n1, n1, n1.Next
	}
	pre.Next = n2
	return dummy.Next
}

func findMiddle(node *ListNode) *ListNode {
	if node.Next == nil {
		return node
	}
	slow, fast := node, node
	for fast.Next != nil && fast.Next.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	ret := slow.Next
	slow.Next = nil
	return ret
}

File: bwu/linked_list/148-sort-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def sortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head
           
        pre, slow, fast = None, head, head
        while fast and fast.next:
            pre = slow
            fast = fast.next.next
            slow = slow.next
            
        pre.next = None
        
        left = self.sortList(head)
        right = self.sortList(slow)
        return self.merge(left, right)

    def merge(self, p, q):
        dummy = ListNode(0)
        curr = dummy
        while p and q:
            if p.val < q.val:
                curr.next = p
                p = p.next
            else:
                curr.next = q
                q = q.next
            curr = curr.next
                
        if p:
            curr.next = p
        if q:
            curr.next = q
            
        return dummy.next
            
